v1.0.0 - MANDATORY

okidokimiki opened the pull request.

okidokimiki commented on marvelgram.xcodeproj/project.pbxproj:
```diff
@@ -15,7 +15,7 @@
 		04C5C1A7284FDDCA0096D272 /* HeroCollectionViewCell.swift in Sources */ = {isa = PBXBuildFile; fileRef = 04C5C1A6284FDDCA0096D272 /* HeroCollectionViewCell.swift */; };
 		04C5C1AB284FEC7A0096D272 /* NetworkManager.swift in Sources */ = {isa = PBXBuildFile; fileRef = 04C5C1AA284FEC7A0096D272 /* NetworkManager.swift */; };
 		04C5C1AE284FED200096D272 /* Hero.swift in Sources */ = {isa = PBXBuildFile; fileRef = 04C5C1AD284FED200096D272 /* Hero.swift */; };
-		04C5C1B1284FEE090096D272 /* HeroViewModel.swift in Sources */ = {isa = PBXBuildFile; fileRef = 04C5C1B0284FEE090096D272 /* HeroViewModel.swift */; };
```
Переименовал `ViewModel` в `Model`, чтобы не сбивать с толку

okidokimiki commented on marvelgram/Modules/Gallery/Views/Cell/HeroImageView.swift:
```diff
@@ -29,19 +32,25 @@ class HeroImageView: UIImageView {
     func loadImageWith(urlString: String) {
         image = nil
-        imageURLString = urlString
+        lastImageURLStringUsedToLoadImage = urlString
```
Переименовал, тк понял, что посмотрев на этот код через некоторое время сразу не вспомнил зачем нужна была эта прочерти. Так вроде лучше читается

bsergiov replied:
так и есть, картинка заниляется для ячейки и формируется исходя из urlString новая картинка, либо из кэша либо из сети

okidokimiki commented on marvelgram/Modules/Gallery/Views/Cell/HeroImageView.swift:
```diff
@@ -11,14 +11,17 @@ class HeroImageView: UIImageView {
-    private var imageURLString: String?
+    private var lastImageURLStringUsedToLoadImage: String?
+    
+    let activityIndicatorView = HeroImageView.makeActivityIndicatorView()
```
Не знаю почему сразу не доавил лоадер на имеджВью пока грузится картинка из интерента

bsergiov replied:
вьюха которая рисует картинку в стартовом состоянии имеет лодер активный и крутящейся, после того как картинка готова к отображение, стопается лоадер и назначается картинка

okidokimiki commented on marvelgram/Modules/Gallery/Views/Cell/HeroImageView.swift:
```diff
@@ -57,22 +66,40 @@ class HeroImageView: UIImageView {
+        activityIndicatorView.startAnimating()
         
-        let task = URLSession.shared.dataTask(with: url) { data, _, _ in
+        let task = URLSession.shared.dataTask(with: url) { resumeDataOrNil, _, errorOrNil in
```
Если будет вопрос зачем переименовал data в resumeDataOrNil и создал errorOrNil, то это сделал это, когда нашел [доку от эпла](https://github.com/brunonagy/ImageLoaderManager/blob/main/ImageLoaderExtension.swift). Пока еще сам не знаю на сколько это правильно и легче ли читается, но решил попробовать так

bsergiov replied:
имена можно давать какие душе угодно, главно что бы через год самому легко вспомнить что за объект прячется под этим именем. И если проект командный то важно что бы имена были понятны для всех участников проекта

okidokimiki commented:
Заметил, что частенько этот код выносит в экстешены, например вот: 
- [ссылка 1](https://github.com/brunonagy/ImageLoaderManager/blob/main/ImageLoaderExtension.swift)
- [ссылка 2](https://github.com/victorpanitz/Swift_ImageLoaderExtension/blob/master/UIImageViewExtension.swift)

okidokimiki commented on marvelgram/Modules/Gallery/Views/Cell/HeroImageView.swift:
```diff
@@ -29,19 +32,25 @@ class HeroImageView: UIImageView {
         }
         
-        makeAndResumeDataTaskWith(urlString: urlString) { [weak self] result in
+        activityIndicatorView.centerXAnchor.constraint(equalTo: centerXAnchor).isActive = true
```
Нормально ли, что я здесь центрирую лоудер? Не очень понимаю как это сделать по-другому

bsergiov replied:
лоадер можно центрировать по картинки
someActivityIndicator.center = someImgView.center

okidokimiki commented on marvelgram/Modules/Gallery/Views/Cell/HeroImageView.swift:
```diff
@@ -57,22 +66,40 @@ class HeroImageView: UIImageView {
+    
+    // MARK: - Creating Subviews
+    
+    static func makeActivityIndicatorView() -> UIActivityIndicatorView {
```
Хотел сделать отдельный класс для этого ActivityIndicatorView, но мне тут нужен `style: .large` и `color = .red`

bsergiov replied:
я бы не стал выносить его в отдельный класс, можно подумать над кастомным вью, который принимает на вход вью и урл.  на выходе рисует эту же вью с лоадером и по завершению загрузки отрисовывает картинку

bsergiov commented:
> @bsergiov, я заметил, что частенько код с загрузкой изображения выносит в экстешены, например вот так:
> 
> * [ссылка 1](https://github.com/brunonagy/ImageLoaderManager/blob/main/ImageLoaderExtension.swift)
> * [ссылка 2](https://github.com/victorpanitz/Swift_ImageLoaderExtension/blob/master/UIImageViewExtension.swift)
> 
> Однако мне не нравится, что в самом фале `UIImageView + Extension` вне экстеншена могу хранится какие-то свойств. Например, тот же кэш и стринги ссылка на картинку `lastImageURLStringUsedToLoadImage`.
> 
> Как вариант, можно создать `class ImageLoad: UIImageView`, внутри папки `Services`, где будет скачиваться картинка, и крутиться лоудер как факт загрузки чего-либо. Однако идея, что нужно либо в экстенщен вынести, либо в нетворкМенеджере качать картинку — не отпускает меня.

'''я стараюсь не допускать глобальных сущностей вне объекта, но такую реализацию не раз встречал в учебных кейсах, на каком то из докладов проскакивала инфа что не позитивно может отражаться на памяти такой подход

okidokimiki committed:
Layout HeroesList screen.

Subviews:
- CollectionView
- NavBarButton
- HeroCell (ReuseView)

Extensions:
-  UICollectionView+ReuseId

okidokimiki committed:
Layout HeroDetails screen.

okidokimiki committed:
Merge pull request #1 from okidokimiki/feature/MRVL-0-Layout

Feature/mrvl 0 layout

okidokimiki committed:
Load & Cache images.

okidokimiki committed:
Save JSON to FileManager.

okidokimiki committed:
Merge pull request #2 from okidokimiki/feature/MRVL-1-Services

Feature/mrvl 1 services

okidokimiki committed:
Open HeroDetails Screen.

okidokimiki committed:
Merge pull request #3 from okidokimiki/feature/MRVL-2-Navigate

Open HeroDetails Screen.
